### PR TITLE
CPP-4991 S5274 Rework rule to cover only optimization inhibiting moves and not all redundant move

### PR DESCRIPTION
**Pull request by tomasz-kaminski-sonarsource**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S5274/cfamily/rule.adoc [hunk 1]
```diff
@@ -1,27 +1,165 @@
...
 
-* in a return statement, if the returned object is a prvalue of the same class type as the function return type
-* in the initialization of a variable, if the initializer expression is a prvalue of the same class type as the variable type
+=== Why is the issue raised if my class does not have a move constructor?
 
-This rule reports an issue when the use of ``++std::move++`` prevents the copy elision from happening.
+A move itself is not performing any object operation, and casting a source to `rvalue`.
+This leads to the constructor and assignment operator that accepts rvalue reference as a parameter,
+also referred to as move constructor and move assignment, to be selected by overload resolution.
+However, in a case when the class does not provide such a constructor,
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:89]:
```suggestion
However, when the class does not provide such a constructor,
```

File: rules/S5274/cfamily/rule.adoc [hunk 2]
```diff
@@ -1,27 +1,165 @@
...
+thus it is not applicable if the parameter is reference to the same type:
+
+[source,cpp]
+----
+void process(A&& sink);
+
+void passArgument() {
+  // No move operation is triggered, as the parameter is a reference to A
+  process(std::move(create())); // Noncompliant, makes changes non-local
+  process(create());            // Complaint
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:123]:
```suggestion
  process(create());            // Compliant
```

File: rules/S5274/cfamily/rule.adoc [hunk 3]
```diff
@@ -1,27 +1,165 @@
...
+[source,cpp]
+----
+class B {
+  // Converting constructor takes object B by value
+  B(A a); 
+};
+void processB(B&& sink);
+
+void passArgument() {
+  processB(create());            // Complaint
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:145]:
```suggestion
  processB(create());            // Compliant
```

File: rules/S5274/cfamily/rule.adoc [hunk 4]
```diff
@@ -1,27 +1,165 @@
...
+  process(std::move(create())); // Noncompliant, makes changes non-local
+  process(create());            // Complaint
+}
+----
+
+Such redundant calls to `std::move` are not inhibiting optimization at this point.
+However, when the `process` function is modified to accept `A` by value,
+it will prevent the compiler from eliminating the move operation altogether.
+To fully benefit from performance benefits change, 
+the programmer would need to review and update all call sides and process functions,
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:131]:
```suggestion
the maintainers would need to review and update all call sites and process functions,
```

File: rules/S5274/cfamily/rule.adoc [hunk 5]
```diff
@@ -1,27 +1,165 @@
...
+  // No move operation is triggered, as the parameter is a reference to A
+  process(std::move(create())); // Noncompliant, makes changes non-local
+  process(create());            // Complaint
+}
+----
+
+Such redundant calls to `std::move` are not inhibiting optimization at this point.
+However, when the `process` function is modified to accept `A` by value,
+it will prevent the compiler from eliminating the move operation altogether.
+To fully benefit from performance benefits change,
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:130]:
```suggestion
To fully benefit from the performance benefits of this change, 
```

File: rules/S5274/cfamily/rule.adoc [hunk 6]
```diff
@@ -1,27 +1,165 @@
...
+----
+
+Such redundant calls to `std::move` are not inhibiting optimization at this point.
+However, when the `process` function is modified to accept `A` by value,
+it will prevent the compiler from eliminating the move operation altogether.
+To fully benefit from performance benefits change, 
+the programmer would need to review and update all call sides and process functions,
+reducing the maintainability of the code.
+
+Moreover, if the parameter is a reference of the type to which the converted object can be converted,
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:134]:
```suggestion
Moreover, if the parameter is a reference to a type to which the converted object can be converted,
```

File: rules/S5274/cfamily/rule.adoc [hunk 7]
```diff
@@ -1,27 +1,165 @@
...
+  processB(create());            // Complaint
+  processB(std::move(create())); // Noncompliant, inhibits copy-elision when initializing constructor parameter
+  // This call is equivalent to:
+  processB(B(std::move(create()))); // Noncompliant, inhibits copy-elision when initializing constructor parameter 
+}
+----
+
+=== Why issues are not raised for all redundant moves?
+
+The requirements from performing an implicit move where relaxed in {cpp}20 and {cpp}23 standards,
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:154]:
```suggestion
The requirements from performing an implicit move were relaxed in {cpp}20 and {cpp}23 standards,
```

File: rules/S5274/cfamily/rule.adoc [hunk 8]
```diff
@@ -1,27 +1,165 @@
...
+with some of them being applied retroactively.
+As a consequence depending on the language and compiler, 
+a call to `std::move` may or not be redundant in the return statement,
+and thus required for the code to be portable accross compilers.
+
+== How to fix it
+
+Remove  call to `std::move` flagged by the rule.
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:162]:
```suggestion
Remove call to `std::move` flagged by the rule.
```

**Review by frederic-tingaud-sonarsource — CHANGES_REQUESTED** [review]

File: rules/S5274/cfamily/rule.adoc [hunk 9]
```diff
@@ -1,27 +1,165 @@
...
 
-If the user writes ``++std::move++`` in one situation that is already handled by the first two cases, it has two drawbacks:
+void asParam(A a);
 
-* It is clumsy, useless code, which makes understanding the code more complex
-* In some cases, it can decrease performances because this can deactivate another optimization of the compiler, named copy elision.
+A returnedFromFunc() {
+  // Move operations need to be performed, and cannot be elided.
+  A a = std::move(create()); // Noncompliant
+  asParam(createA());        // Noncompliant
```
[the remark this answers is not in the document]
Comment (tomasz-kaminski-sonarsource) [reply rules/S5274/cfamily/rule.adoc:49]: Yes, added.

File: rules/S5274/cfamily/rule.adoc [hunk 10]
```diff
@@ -1,27 +1,165 @@
...
+  return A();                // Noncompliant
```
[the remark this answers is not in the document]
Comment (tomasz-kaminski-sonarsource) [reply rules/S5274/cfamily/rule.adoc:50]: Yes

File: rules/S5274/cfamily/rule.adoc [hunk 11]
```diff
@@ -1,29 +1,174 @@
...
+  OnlyCopyable c1 = std::move(create()); // Noncompliant
+  // Copy elision eliminates invocation of the copy constructor
+  OnlyCopyable c2 = create(); // Compliant
+}
+----
+
+=== Why is an issue raised when passing an argument to a reference parameter?
 
+The copy elision optimization applies only if a new value is produced from the source,
+thus it is not applicable if the parameter is reference to the same type:
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:119]:
```suggestion
The copy elision optimization happens only if a new value is produced from the source,
not if the parameter is a reference to the same type:
```

File: rules/S5274/cfamily/rule.adoc [hunk 12]
```diff
@@ -1,29 +1,174 @@
...
+[source,cpp]
+----
+void process(A&& sink);
 
-* The object is a temporary object with no name, and if it can't be named, it can't be used
-* The object is used in some specific places, such as a return statement
-* The user explicitly promises to the compiler that they won't care for the object's current value any longer. They do so by using the specific cast operation named ``++std::move++``.
+void passArgument() {
+  // No move operation is triggered, as the parameter is a reference to A
+  process(std::move(create())); // Noncompliant, makes changes non-local
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:127]: The sample already mentions what will be explained in the next paragraph. I think that is a bit confusing. Perhaps it should be moved after the explanation about maintainability?

File: rules/S5274/cfamily/rule.adoc [hunk 13]
```diff
@@ -1,29 +1,174 @@
...
+However, when the `process` function is modified to accept `A` by value,
+it will prevent the compiler from eliminating the move operation altogether.
+To fully benefit from the performance impact of this change, 
+the maintainers would need to review and update all call sites and process functions,
+reducing the maintainability of the code.
 
-* It is clumsy, useless code, which makes understanding the code more complex
-* In some cases, it can decrease performances because this can deactivate another optimization of the compiler, named copy elision.
+Moreover, if the parameter is a reference to a type to which the converted object can be converted,
+then copy elision may still happen when initializing the parameter of the converting constructor.
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:140]:
```suggestion
Moreover, if the parameter is a reference to a type to which the argument is converted,
then copy elision may still happen when calling the converting constructor.
```

File: rules/S5274/cfamily/rule.adoc [hunk 14]
```diff
@@ -1,29 +1,174 @@
...
+}
+----
 
-When copy elision occurs, the object is neither copied nor moved (even if the copy/move constructors have side effects). The two objects are collapsed into only one memory location. The moment when copy elision occurs is compiler-dependent but is mandatory in the following cases:
+=== Why issues are not raised for all redundant moves?
 
+The requirements from performing an implicit move were relaxed in {cpp}20 and {cpp}23 standards,
+with some of them being applied retroactively.
+As a consequence depending on the language and compiler, 
+a call to `std::move` may or not be redundant in the return statement,
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:162]:
```suggestion
As a consequence depending on the standard and compiler versions, 
a call to `std::move` may or may not be redundant in the return statement,
```

File: rules/S5274/cfamily/rule.adoc [hunk 15]
```diff
@@ -1,29 +1,174 @@
...
+and thus required for the code to be portable accross compilers.
 
-* in a return statement, if the returned object is a prvalue of the same class type as the function return type
-* in the initialization of a variable, if the initializer expression is a prvalue of the same class type as the variable type
+== How to fix it
 
-This rule reports an issue when the use of ``++std::move++`` prevents the copy elision from happening.
+Remove call to `std::move` flagged by the rule.
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:167]:
```suggestion
Remove the call to `std::move` flagged by the rule.
```

File: rules/S5274/cfamily/rule.adoc [hunk 16]
```diff
@@ -48,9 +191,35 @@ A getA();
...
+
+
+==== Compliant solution
+
+[source,cpp,diff-id=2,diff-type=compliant]
+----
+class A {};
+A getA();
+
+A f() {
```
Comment (frederic-tingaud-sonarsource) [inline rules/S5274/cfamily/rule.adoc:220]: The function doesn't return `A` anymore.

**Review by frederic-tingaud-sonarsource — APPROVED** [review]

File: rules/S5274/cfamily/rule.adoc (hunk 12 above)
Comment (tomasz-kaminski-sonarsource) [reply rules/S5274/cfamily/rule.adoc:127]: The process here is taken by reference, so this is an illustration for the previous paragraph. Yes, the Noncomplain have a explanation referring to what is later. I can remove it, if you would consider that cleaner.

File: rules/S5274/cfamily/rule.adoc (hunk 16 above)
Comment (tomasz-kaminski-sonarsource) [reply rules/S5274/cfamily/rule.adoc]: Fixed

File: rules/S5274/cfamily/rule.adoc (hunk 12 above)
Comment (tomasz-kaminski-sonarsource) [reply rules/S5274/cfamily/rule.adoc]: I would find it more confusing if that example would be after a paragraph that explained the consequences of changing something to be by value. And the section title already says that the issue is raised.

**Pull request merged**